common/src/main/java/com/wynntils/screens/base/widgets/TextInputBoxWidget.java

```diff
@@ -85,88 +98,221 @@ public void renderWidget(PoseStack poseStack, int mouseX, int mouseY, float part
-                    && stringPosition < this.textBoxInput.length()) {
-                builder.append(textBoxInput.charAt(stringPosition));
+        // Now reverse so it's actually to the left
+        builder.reverse();
+        builder.append(cursorChar);
 
-                stringPosition++;
-            }
+        // Now append to the right of the cursor
+        stringPosition = cursorPosition;
+        while (font.width(builder.toString()) < maxTextWidth - cursorWidth
+                && stringPosition < this.textBoxInput.length()) {
+            builder.append(textBoxInput.charAt(stringPosition));
 
-            renderedText = builder.toString();
+            stringPosition++;
         }
 
-        return renderedText;
+        return new Pair<>(builder.toString(), startingAt);
     }
 
     @Override
     public boolean mouseClicked(double mouseX, double mouseY, int button) {
         McUtils.playSound(SoundEvents.UI_BUTTON_CLICK.value());
 
         if (this.isHovered) {
+            setCursorAndHighlightPositions(getIndexAtPosition(mouseX));
+            isDragging = true;
             textboxScreen.setFocusedTextInput(this);
             return true;
         } else {
+            setCursorAndHighlightPositions(cursorPosition); // remove highlights when clicking off
             textboxScreen.setFocusedTextInput(null);
         }
 
         return false;
     }
 
+    @Override
+    public boolean mouseReleased(double mouseX, double mouseY, int button) {
+        if (isDragging) {
+            isDragging = false;
+            setCursorPosition(getIndexAtPosition(mouseX));
+        }
+        return super.mouseReleased(mouseX, mouseY, button);
+    }
+
+    @Override
+    public boolean mouseDragged(double mouseX, double mouseY, int button, double dragX, double dragY) {
+        if (isDragging) {
+            setCursorPosition(getIndexAtPosition(mouseX));
+        }
+        return super.mouseDragged(mouseX, mouseY, button, dragX, dragY);
+    }
+
+    protected int getIndexAtPosition(double mouseX) {
+        mouseX -= this.getX(); // mouseX is actually just the x position of the mouse relative to the screen, not the textbox
+        Font f = FontRenderer.getInstance().getFont();
+        Pair<String, Integer> renderedTextDetails = getRenderedText(maxTextWidth, false);
+        String renderedText = renderedTextDetails.a();
+
+        // Width so far at index i
+        List<Float> widths = new ArrayList<>();
+        for (int i = 0; i < renderedText.length(); i++) {
+            // we are using stringWidth here because we need precision; if we use width, it will round to the nearest
+            // integer, which will cause strange behaviour when clicking on letters
+            widths.add(f.getSplitter().stringWidth(renderedText.substring(0, i)));
+        }
+
+        // get nearest width that's in the map
+        // FIXME: this is probably really slow and bad, but I am just a first year cs student and I have not taken
```

Comment:
We have `com.wynntils.utils.mc.RenderedStringUtils#getMaxFittingText` which does all the work you need here. You need to just make the `...` optional, or create a new method.

Reply:
I think `getMaxFittingText(renderedText, mouseX, font).length() - 1 + renderedTextDetails.b()`, but I might not understood the code fully.

Reply:
(Same as mentioned in Discord)

Can't use this method, as it rounds to the leftmost character when between characters